Open.h

```diff
@@ -3,16 +3,206 @@
 #include "HashTable.h"
 #include "Product.h"
 #include <iostream>
-#include <map>
+#include <vector>
+#include <cmath>
+#include <string>
 
 template<typename Key, typename Value>
 class OpenMap: public HashTable<Key, Value> {
-    public:
-        OpenMap() = default;
+private:
+    double MAX_LOAD_FACTOR = 0.7;
 
-        Value get(Key k) = 0;
+    struct KeyValuePair{
+        KeyValuePair(Key k, Value v) {
+            key = k;
+            value = v;
+            valid = true;
+        }
+        KeyValuePair(){
+            key = Key{};
+            value = Value{};
+            valid = false;
+        }
+        Key key;
+        Value value;
+        bool valid;
+    };
 
-        void put(Key k, Value v) = 0;
+    //this is the operation done on any index that collides
+    uint64_t iterate(uint64_t index){
+        return index*index;
+    }
 
-        void remove(Key k) = 0;
+    std::vector<KeyValuePair>* hashmap;
+    std::vector<bool>* significancemap;
+
+    std::hash<Key> hasher;
+
+    uint64_t hash(Key k){
+        return hasher(k)%hashmapSize;
+    }
+
+    uint64_t hashmapSize = 4;
+
+    uint64_t numPairs = 0;
+    uint64_t numSignificant = 0;
+
+    void doubleSize(){
```

Comment:
Is there a way you can write this that uses public APIs?

Reply:
I think I could, but repeatedly calling the put() function is a lot less efficient than this. Because I know I won't have any duplicate values and I won't need to double size, this function can go a lot faster.

Reply:
Penny wise, pound foolish.

Use public APIs when you can because they will be correct.  Is resizing your hash table something you are going to do in a tight loop?  So why optimize it to that extent?